File: lambda/login/module/auth/login_refresh_token.go
package auth

import (
	"daita_module_login/component/common"
	"daita_module_login/component/hash"
	usermodel "daita_module_login/dynamoDB/user/model"
	"encoding/json"
	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/cognitoidentityprovider"
	"net/http"
	"time"
)

func (app Application) authInputRefreshToken(params map[string]*string) *cognitoidentityprovider.InitiateAuthInput {
	authenCognito := &cognitoidentityprovider.InitiateAuthInput{}
	authenCognito.AuthFlow = aws.String("REFRESH_TOKEN_AUTH")
	authenCognito.ClientId = aws.String(app.Configure.ClientPooID)
	authenCognito.AuthParameters = params
	return authenCognito
}

func (app Application) loginRefreshToken(params map[string]*string) (usermodel.UserRefreshTokenResponse, error) {

	authInput := app.authInputRefreshToken(params)
	res, err := app.Coginto.InitiateAuth(authInput)
	if err != nil {

		return usermodel.UserRefreshTokenResponse{}, err
	}
	IDtoken := *res.AuthenticationResult.IdToken
	sess, _ := session.NewSession(&aws.Config{
		Region: aws.String(common.REGION),
	})
	credentialUser, identityId, err := app.getCredentialsForIdentity(&IDtoken, sess)
	userAuthResponse := usermodel.UserRefreshTokenResponse{
		AccessKey:                *credentialUser.AccessKeyId,
		SecretKey:                *credentialUser.SecretKey,
		SessionKey:               *credentialUser.SessionToken,
		Token:                    *res.AuthenticationResult.AccessToken,
		IdentityID:               *identityId,
		CredentialTokenExpiresIn: ((*credentialUser.Expiration).UnixNano()) / int64(time.Millisecond),
		IDToken:                  IDtoken,
		TokenExpiresIn:           time.Now().AddDate(0, 0, 1).UnixNano() / int64(time.Millisecond),
	}
	return userAuthResponse, nil
}
func (app Application) refreshTokenHandler(event events.APIGatewayProxyRequest, headers map[string]string) (string, int, map[string]string, map[string]interface{}, error) {
	data := map[string]interface{}{}
	user := usermodel.UserRefreshToken{}
	err := json.Unmarshal([]byte(event.Body), &user)
	if err != nil {
		message := common.MessageUnmarshalInputJson
		return message, http.StatusOK, headers, data, common.ErrUnmarshalInputJson
	}
	params := map[string]*string{
		"REFRESH_TOKEN": aws.String(user.RefreshToken),
		"SECRET_HASH":   aws.String(hash.CreateSecretHash(app.Configure.ClientPoolSecret, user.UserName, app.Configure.ClientPooID)),
	}
	userResponse, err := app.loginRefreshToken(params)
	if err != nil {
		message := common.MessageRefreshTokenError
		return message, http.StatusOK, headers, data, common.ErrAuthenFailed
	}
	message := common.MessageRefreshTokenSuccessfully
	inrec, _ := json.Marshal(userResponse)
	json.Unmarshal(inrec, &data)
	return message, http.StatusOK, headers, data, nil
}

// implement login refresh token
func (app Application) ImpLoginRefreshToken(event events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	headers := map[string]string{"Access-Control-Allow-Credentials": "true",
		"Access-Control-Allow-Methods": "GET,HEAD,OPTIONS,POST,PUT",
	}
	message := ""
	statusCode := http.StatusOK
	err := common.ErrInit
	data := map[string]interface{}{}
	message, statusCode, headers, data, err = app.refreshTokenHandler(event, headers)
	return common.GenerateResponseBody(message, statusCode, headers, data, err, []string{}), nil
}

File: core_service/aws_lambda/project/code/login_refresh_token.py
import os
import json
import logging
from datetime import datetime
from http import HTTPStatus

import boto3

from config import *
from error import *
from response import generate_response, error_response
from utils import create_secret_hash, aws_get_identity_id


logger = logging.getLogger(__name__)
cog_provider_client = boto3.client('cognito-idp')
cog_identity_client = boto3.client('cognito-identity')
ACCESS_TOKEN_EXPIRATION = 24 * 60 * 60
RESPONSE_HEADER = {
    "Access-Control-Allow-Credentials": "true",
	"Access-Control-Allow-Methods": "GET, HEAD, OPTIONS, POST, PUT",
}


@error_response
def lambda_handler(event, context):
    '''
    https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-concepts.html#gettingstarted-concepts-event
    https://docs.aws.amazon.com/lambda/latest/dg/python-context.html
    '''
    # get params REFRESH_TOKEN passed by FE in event body
    try:
        username = event["username"]
        refresh_token = event["refresh_token"]
    except Exception as exc:
        raise Exception(MessageUnmarshalInputJson) from exc

    try:
        # init Cognito InitiateAuth
        authenticated = cog_provider_client.initiate_auth(
                AuthFlow="REFRESH_TOKEN_AUTH",
                AuthParameters={
                    "REFRESH_TOKEN": refresh_token,
                    "SECRET_HASH": create_secret_hash(
                        clientSecret="",
                        username=username,
                        clientPoolID=CLIENTPOOLID
                    )
                },
                ClientId=CLIENTPOOLID,
            )

        id_token = authenticated["AuthenticationResult"]["IdToken"]
        IdentityId = aws_get_identity_id(id_token)

        # call Cognito getCredentialsForIdentity
        identity = cog_identity_client.get_credentials_for_identity(
                IdentityId=IdentityId,
                Logins={
                   f'cognito-idp.{cog_identity_client.meta.region_name}.amazonaws.com/{os.environ["USER_POOL_ID"]}': id_token
                }
            )
    except Exception as exc:
        raise Exception(MessageRefreshTokenError) from exc

    # reformat
    user_data = {
        "access_key": identity["Credentials"]["AccessKeyId"],
        "secret_key": identity["Credentials"]["SecretKey"],
        "session_key": identity["Credentials"]["SessionToken"],
        "token": authenticated["AuthenticationResult"]["AccessToken"],
        "identity_id": identity["IdentityId"],
        "credential_token_expires_in": (identity["Credentials"]["Expiration"].timestamp())*1000, # expire time in seconds
        "id_token": id_token,
        "token_expires_in":float(int((datetime.now().timestamp() + ACCESS_TOKEN_EXPIRATION)*1000)),
        "username": username
    }

    # return response
    return generate_response(
        message=MessageRefreshTokenSuccessfully,
        status_code=HTTPStatus.OK,
        headers=RESPONSE_HEADER,
        data=user_data,
    )
